packages/devextreme/js/__internal/ui/map/m_provider.dynamic.azure.ts

```diff
@@ -0,0 +1,499 @@
+
+    const $content = $('<div>').text(options.text).addClass(MAP_MARKER_TOOLTIP_CLASS);
+    const popup = new atlas.Popup({
+      content: $content[0],
+      position: location,
+      pixelOffset: [0, -30],
+    });
+
+    this._map.popups.add(popup);
+
+    if (options.visible) {
+      popup.open();
+    }
+
+    return popup;
+  },
+
+  _destroyMarker(marker) {
+    this._map.markers.remove(marker.marker);
+    if (marker.popup) {
+      this._map.popups.remove(marker.popup);
+    }
+    if (marker.handler) {
+      this._map.events.remove(marker.handler);
+    }
+  },
+
+  _renderRoute(options) {
+    return Promise.all(
+      map(options.locations, (point) => this._resolveLocation(point)),
+    ).then((locations) => new Promise((resolve) => {
+      const routeColor = new Color(options.color || this._defaultRouteColor()).toHex();
+      const routeOpacity = options.opacity || this._defaultRouteOpacity();
+      const queryCoordinates = locations.map((location) => `${location[1]},${location[0]}`);
+      const query = queryCoordinates.join(':');
+      const routeType = this._movementMode(options.mode);
+      const isWalkingType = routeType === this._movementMode('walking');
+      const searchUrl = `${AZURE_LINK}route/directions/json?subscription-key=${this._keyOption('azure')}&api-version=1.0&query=${query}&travelMode=${routeType}`;
+
+      ajax.sendRequest({
+        url: CUSTOM_URL ?? searchUrl,
+        dataType: 'json',
+      }).then((result) => {
+        if (result?.routes && result.routes.length > 0) {
+          const route = result.routes[0];
+          const routeCoordinates = route.legs
+            .flatMap((leg) => leg.points.map((point) => [point.longitude, point.latitude]));
+          const dataSource = new atlas.source.DataSource();
+
+          dataSource.add(new atlas.data.Feature(new atlas.data.LineString(routeCoordinates), {}));
+
+          const lineLayerConfig: {
+            strokeColor: string;
+            strokeOpacity: number;
+            strokeWidth: number;
+            strokeDashArray?: number[];
+          } = {
+            strokeColor: routeColor,
+            strokeOpacity: routeOpacity,
+            strokeWidth: options.weight || this._defaultRouteWeight(),
+          };
+
+          if (isWalkingType) {
+            lineLayerConfig.strokeDashArray = [1, 1];
+          }
+
+          const lineLayer = new atlas.layer.LineLayer(dataSource, null, lineLayerConfig);
+
+          this._map.sources.add(dataSource);
+          this._map.layers.add(lineLayer);
+
+          const bounds = atlas.data.BoundingBox.fromPositions(routeCoordinates);
+          if (this._option('autoAdjust')) {
+            this._map.setCamera({
+              bounds,
+              padding: 50,
+            });
+          }
+
+          resolve({
+            instance: { dataSource, lineLayer },
+            northEast: [bounds[2], bounds[3]],
+            southWest: [bounds[0], bounds[1]],
+          });
+        }
```

Comment:
[refactoring]
Cannot we split this code? For example implement separate `_createLineLayer` method, or maybe smth else
Not it's difficult to read

Reply:
I removed logic related to stroke dash adding for walking route. Now there is no excess if and no separate LineLayer config, I think there is no need to move it to a separate function now